src/main/java/org/truffleruby/core/hash/library/CompactHashStore.java

```diff
@@ -0,0 +1,416 @@
+        dest.compareByIdentity = hash.compareByIdentity;
+        dest.defaultBlock = hash.defaultBlock;
+        dest.defaultValue = hash.defaultValue;
+    }
+    @ExportMessage
+    protected RubyArray shift(RubyHash hash,
+                              @Shared @Cached HashingNodes.ToHash hashFunction,
+                              @Shared @Cached GetHashPosInIndexNode getHashPos,
+                              @CachedLibrary("this") HashStoreLibrary self) {
+        int lastKeyPos = kvStoreInsertionPos - 2;
+        while (lastKeyPos > 0 &&
+               kvStore[lastKeyPos] == null) {
+            lastKeyPos -= 2;
+        }
+        if (lastKeyPos < 0) return null;
+
+        Object key = kvStore[lastKeyPos];
+        Object val = kvStore[lastKeyPos+1];
+
+        int keyHash = hashFunction.execute(key,hash.compareByIdentity);
+        int hashPos = getHashPos.execute(keyHash,index);
+
+        index[hashPos+1] = INDEX_SLOT_DELETED_MARKER;
+        kvStore[lastKeyPos] = null;
+        kvStore[lastKeyPos+1] = null;
+        hash.size--;
+
+        return ArrayHelpers.createArray(
+                RubyContext.get(self),
+                RubyLanguage.get(self),
+                new Object[]{key,val});
+    }
+    @ExportMessage
+    protected void rehash(RubyHash hash,
+                          @Cached @Shared HashingNodes.ToHash hashFunction) {
+        for (int i = 0; i < index.length; i++) {
+            int kvPos = index[i+1];
+            boolean slotFilled = kvPos > INDEX_SLOT_EMPTY_MARKER;
+            if (slotFilled) {
+                kvPos--;
+
+                Object key = kvStore[kvPos];
+                index[i] = hashFunction.execute(key,hash.compareByIdentity);
+            }
+        }
+    }
+    @ExportMessage
+    protected boolean verify(RubyHash hash) {
+        assert kvStoreInsertionPos > 0  && kvStoreInsertionPos < kvStore.length;
+        assert kvStoreInsertionPos % 2 == 0;
+        assert kvStoreInsertionPos >= hash.size; //because deletes only decrease hash.size
+        assert ((float)(2 * hash.size))/index.length <= THRESHOLD_LOAD_FACTOR_FOR_INDEX_REBUILD;
+        return true;
+    }
+
+    private void insertIntoIndex(int keyHash, int kvPos) {
+        int pos = Math.abs(keyHash % index.length) & 0xFF_FF_FF_FE;
+
+        boolean slotFilled = index[pos+1] > INDEX_SLOT_EMPTY_MARKER;
+        while (slotFilled) {
+            pos = (pos + 2) % index.length;
+            slotFilled = index[pos + 1] > INDEX_SLOT_EMPTY_MARKER;
+        }
+
+        index[pos] = keyHash;
+        index[pos+1] = kvPos;
+    }
+
+    private void insertIntoKv(Object key, Object value) {
+        kvStore[kvStoreInsertionPos] = key;
+        kvStore[kvStoreInsertionPos+1] = value;
+        kvStoreInsertionPos += 2;
+    }
+
+    private void resizeIndexIfNeeded(int size) {
+        float loadFactor = ((float) (2 * size))/index.length;
+        if (loadFactor < THRESHOLD_LOAD_FACTOR_FOR_INDEX_REBUILD) return;
```

Comment:
I think we should avoid any float/double computation on the fast, that's pretty expensive.
Instead we could store an `int` to know at which number of entries we should resize.

Reply:
Hmm, I understand your point about float div. 

I'm not sure I understand the "we could store an int" bit, resizing doesn't happen at a fixed known number of entries, it happens whenever the ration of slots used (hash.size) to the total number of slots (index.length/2). At least, you would need an int[] lookup table to replace that, not a single int.

Alternatively, I found that SO question https://stackoverflow.com/questions/20788793/c-how-to-calculate-a-percentageperthousands-without-floating-point-precision, where people found a way to encode a ration using ints (25.5 --> 255, etc...). The caclulation is (100*a + b/2)/b, so just a multiplication, a shift (/2), and integer divsion. Is that efficient enough ?

Reply:
> it happens whenever the ratio of slots used (hash.size) to the total number of slots (index.length/2) exceeds a threshold

So whenever you resize, you compute `threshold = loadFactor * index.length / 2`. And whenever we add an entry via e.g. Hash#[]= we'd only check `if (hash.size == threshold)`